pkg/reconciler/topology/partitionset/partitionset_reconcile.go

```diff
@@ -0,0 +1,208 @@
+				conditionsv1alpha1.ConditionSeverityError,
+				"",
+			)
+			// No need to requeue if the selector is not valid
+			return nil
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionSetValid)
+
+	shards, err := c.listShards(labelSelector)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
+		)
+		return err
+	}
+
+	oldPartitions, err := c.getPartitionsByPartitionSet(partitionSet)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return err
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, partitionSet.Spec.ShardSelector.MatchLabels)
+	} else {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, nil)
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
+	var placeHolder struct{}
+	newMatchExpressions := []metav1.LabelSelectorRequirement{}
+	if partitionSet.Spec.ShardSelector != nil {
+		newMatchExpressions = partitionSet.Spec.ShardSelector.MatchExpressions
+	}
+	for _, oldPartition := range oldPartitions {
+		// MatchExpressions need to be the same
+		oldMatchExpressions := []metav1.LabelSelectorRequirement{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil && !apierrors.IsNotFound(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+			continue
+		}
+
+		// MatchLabels need to be the same
+		partitionKey := ""
+		oldMatchLabels := map[string]string{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchLabels = oldPartition.Spec.Selector.MatchLabels
+		}
+
+		// Sorting the keys of the old partition for consistent comparison
+		sortedOldKeys := make([]string, len(oldMatchLabels))
+
+		i := 0
+		for k := range oldMatchLabels {
+			sortedOldKeys[i] = k
+			i++
+		}
+		sort.Strings(sortedOldKeys)
+		for _, key := range sortedOldKeys {
+			partitionKey = partitionKey + "+" + key + "=" + oldMatchLabels[key]
+		}
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
```

Comment:
We usually put `logger.V(2)` before every client call

Reply:
added